src/ipld/util.rs

```diff
@@ -437,3 +438,259 @@ impl<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin> Stream for ChainStream<
         }
     }
 }
+
+enum UnorderedTask {
+    Cid(Cid),
+}
+
+pin_project! {
+    pub struct UnorderedChainStream<DB, T> {
+        tipset_iter: T,
+        db: Arc<DB>,
+        seen: Arc<Mutex<CidHashSet>>,
+        worker_handle: JoinHandle<anyhow::Result<()>>,
+        block_receiver: kanal::Receiver<anyhow::Result<Block>>,
+        extract_sender: kanal::Sender<Cid>,
+        stateroot_limit: ChainEpoch,
+        queue: Vec<UnorderedTask>,
+        fail_on_dead_links: bool,
+    }
+}
+
+impl<DB, T> UnorderedChainStream<DB, T> {
+    pub fn into_seen(self) -> CidHashSet {
+        match Arc::try_unwrap(self.seen) {
+            Ok(v) => v.into_inner(),
+            Err(v) => v.lock().clone(),
+        }
+    }
+}
+
+/// Stream all blocks that are reachable before the `stateroot_limit` epoch in an unordered fashion.
+/// After this limit, only block headers are streamed. Any dead links are reported as errors.
+///
+/// # Arguments
+///
+/// * `db` - A database that implements [`Blockstore`] interface.
+/// * `tipset_iter` - An iterator of [`Tipset`], descending order `$child -> $parent`.
+/// * `stateroot_limit` - An epoch that signifies how far back we need to inspect tipsets, in-depth.
+/// This has to be pre-calculated using this formula: `$cur_epoch - $depth`, where `$depth` is the
+/// number of `[`Tipset`]` that needs inspection.
+#[allow(dead_code)]
+pub fn unordered_stream_chain<
+    DB: Blockstore + Sync + Send + 'static,
+    T: Iterator<Item = Tipset> + Unpin + Send + 'static,
+>(
+    db: Arc<DB>,
+    tipset_iter: T,
+    stateroot_limit: ChainEpoch,
+) -> UnorderedChainStream<DB, T> {
+    let (sender, receiver) = kanal::bounded(2048);
```

Comment:
Wondering how did you pick this `2048` value? Does the value has any impact on performances or memory usage?

Reply:
It's somewhat arbitrary, 2048 worked well on my machine and memory usage for 2048 items is negligible. I'll introduce a constant to be more transparent.